### DISCUSSION TITLE
Different Embedding Size

### DISCUSSION THREAD

**Discussion by mprithiv** [pyg-team/pytorch_geometric#3944] [Q&A]
Hi

This is the HeteroData object I have
`HeteroData( buyerId={ num_nodes=7813, x=[7813, 1152] }, sellerId={ num_nodes=884, x=[884, 387] }, (buyerId, txn_amount, sellerId)={ edge_index=[2, 335959] }, (buyerId, txn_date, sellerId)={ edge_index=[2, 335959] } )`

I have a different embedding size for both the nodes and the edges. I am trying to understand how important it is to have same embedding size for both the nodes and the edges and impact of padding/shrinking the embedding to make it similar for all the nodes

The embeddings are created using the SentenceTransformer for alpha-numeric and LabelEncoder for numeric
```python
class SequenceEncoder(object):
    def __init__(self, model_name="all-MiniLM-L6-v2", device=None):
        self.device = device
        self.model = SentenceTransformer(model_name, device=None)
    @torch.no_grad()
    def __call__(self, df):
        #cars = set(g for col in df.values)
        x = self.model.encode(df.values, show_progress_bar=True, convert_to_tensor=True)
        return x.cpu()
```
and
```python
class LabelEncoder(object):
    def __init__(self, dtype=None):
        self.dtype = dtype
    #@torch.no_grad()
    def __call__(self, df):
        #ys = pd.read_csv(df)['cbd_total_3months']
        return torch.from_numpy(df.values).view(-1,1).to(self.dtype)#, show_progress_bar=True, convert_to_tensor=True)`
```
How do I set an embedding size while preparing the Data object? Is the different embedding size more of a problem in creating NodeSplit

**Comment by rusty1s** [comment]
Can you clarify on what's the concrete issue? In general, `HeteroData` and our heterogeneous GNNs (such as `to_hetero()`) are designed to be fully able to operate on differently sized embeddings, *e.g.*, via lazy initialization. [Our tutorial](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html) covers more information.

**Reply by mprithiv** [reply]
Hi. Thanks for your reply. As you might remember from my other discussions, my HeteroData object has different embedding sizes.

`HeteroData(
  cardholder={
    num_nodes=7813,
    x=[7813, 1152]
  },
  merchant={
    num_nodes=884,
    x=[884, 387]
  },
  (cardholder, txn_amount, merchant)={ edge_index=[2, 335959] },
  (cardholder, dw_process_date, merchant)={ edge_index=[2, 335959] }
)`

1. When doing to_homogeneous() on the HeteroData with different embedding sizes, I see the `x` on `buyer` and `seller` add up to `7813+884=8697` but then the `x` parameter becomes one-dimensional. I mean, I do not see the node features embedding `Data(node_type=[8697], edge_index=[2, 671918], edge_type=[671918])`
2. As there is no interaction/connections between the `buyer` nodes, for training the node embeddings unsupervised, is HeteroData even a correct representation or rather consider it as a simple bipartite graph
3. I am looking to get the neighours or a similarty of a node - For example, given a node ID, I would like to find the buyers who has done transactions with the same or similar merchants based on some attributes like location, amount, etc. If you could point me out to some relevant examples, would be helpful.

**Reply by rusty1s** [reply]
1. Yes, `to_homogeneous()` will drop all attributes whose feature dimensions do not match across all node types. There is no way for as to merge `data['cardholder'].x` and `data['merchant'].x` into a single feature matrix. I personally think you do not necessarily need to convert your data into a homogeneous graph. Treating it as heterogeneous (even in the model) makes perfect sense. If you still want to use `to_homogeneous()`, you can apply a `Linear` layer on `data['cardholder'].x` and `data['merchant'].x` inside your model, and convert with `to_homogeneous()` afterwards:
```python
data = copy.copy(data)
for node_type in data.node_types:
    data[node_type].x = self.lin_dict[node_type](data[node_type].x)
data = data.to_homogeneous()
```
2. Yes, this looks absolutely correct to me. A bipartite graph is the most simple heterogeneous graph one can think of.
3. Interesting. Spontaneously, I can't think of any reference papers/examples on this. Can you share some information about how you would want to train such a model in an unsupervised fashion in the first place?

**Reply by mprithiv** [reply]
Based on the heterodata above, I would like to do some sampling. I am looking to build an embedding model that takes in the buyer data(attributes: id, bank, etc), seller data(attributes: id, location, type, etc) with two edge types(amount and date). I dont have an explicit attribute I can consider as ground truth. What I am planning to do is

1. Do weighted sampling - get the neighbourhood nodes based on the higher amount transacted 
2. Get this neighbourhood list and use that as a ground truth label

Is there any way to do the weighted sampling with PyG? Also, would you please advise if it is better to consider amount/date as an edge vs 'deals' as an edge with amount/date as an edge attribute? Also, would it be useful to embed both edge and nodes together considering my case? If so, could you point me out to the right guide please.

`out = model(data.x, data.edge_index)` when I have a heterodata with two edge_indices, how do I take both edge_indices into consideration (or should I create one edge_index by creating an edge as 'deals' and keeping amount/date for the edge attributes)

**Reply by rusty1s** [reply]
Weighted sampling is not yet supported, I'm sorry :(

I think in your case it makes more sense to have a single edge type with edge features attached to it (like amount, date). This requires that you utilize a GNN that can incorporate edge features.

You can get both edge indices as a dict by running `data.edge_index_dict`.

**Reply by mprithiv** [reply]
Hi. I was using the [sageconv from here](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html). Does it support incorporating edge features?

On a different note, I have two node types (buyer and seller). The node classification might happen within buyer but there might be a link classification or subgraph generation in the future. So basically, how can i ensure that both the node types(buyer and seller) are encoded along with the edge information (like "Attention Mechanism - Jointly Learning to Align and Translate in NLP"), jointly embed the multiple node and edge information?

**Reply by rusty1s** [reply]
`SAGEConv` does not support edge features, [Here](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html) is an overview of all GNN operators that support them. All operators that support multi-dimensional edge features should guarantee that edge information is incorporated into the resulting node embeddings.

**Reply by mprithiv** [reply]
Is there an example implementation for TransformerConv

**Reply by rusty1s** [reply]
Not yet, but we are trying to integrate one with this issue https://github.com/pyg-team/pytorch_geometric/issues/4040.
